Zend/zend_compile.c

```diff
@@ -7189,12 +7189,71 @@ void zend_compile_class_decl(znode *result, zend_ast *ast, zend_bool toplevel) /
 			zend_error_noreturn(E_COMPILE_ERROR, "Clone method %s::%s() cannot be static",
 				ZSTR_VAL(ce->name), ZSTR_VAL(ce->clone->common.function_name));
 		} else if (ce->clone->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
-			zend_error_noreturn(E_COMPILE_ERROR,
-				"Clone method %s::%s() cannot declare a return type",
-				ZSTR_VAL(ce->name), ZSTR_VAL(ce->clone->common.function_name));
+			if (ZEND_TYPE_FULL_MASK(ce->clone->common.arg_info[-1].type) != MAY_BE_VOID) {
+				zend_error_noreturn(E_COMPILE_ERROR, "%s::__clone(): Return type must be void when declared", ZSTR_VAL(ce->name));
+			}
+		}
+	}
+
+	if (ce->__set && ce->__set->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		if (ZEND_TYPE_FULL_MASK(ce->__set->common.arg_info[-1].type) != MAY_BE_VOID) {
+			zend_error_noreturn(E_COMPILE_ERROR, "%s::__set(): Return type must be void when declared", ZSTR_VAL(ce->name));
+		}
+	}
+
+	if (ce->__unset && ce->__unset->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		if (ZEND_TYPE_FULL_MASK(ce->__unset->common.arg_info[-1].type) != MAY_BE_VOID) {
+			zend_error_noreturn(E_COMPILE_ERROR, "%s::__unset(): Return type must be void when declared", ZSTR_VAL(ce->name));
+		}
+	}
+
+	if (ce->__isset && ce->__isset->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE) {
+		 if (ZEND_TYPE_FULL_MASK(ce->__isset->common.arg_info[-1].type) != MAY_BE_BOOL) {
```

Comment:
This check will allow `: bool|stdClass`.

Reply:
@nikic Which scenario? I've tested declaring a 

```php
class ValidMagicMethods {
	public function __isset(string $name): bool|stdClass {}
}
```

and also extending from this class as

```php
class ValidMagicMethods {
	public function __isset(string $name): bool {}
}

class NarrowedReturnType extends ValidMagicMethods {
	public function __isset(string $name): bool|stdClass {}
}
```

And both failed with

```
Fatal error: ValidMagicMethods::__isset(): Return type must be bool when declared in /Users/gcaruso/Workspaces/PHP/php-src/Zend/tests/magic_methods_inheritance_rules.php on line 30

```

Reply:
Oh right, my bad.